Add Rust CPD

juliapaluch opened the pull request:
## Describe the PR

This PR adds CPD support for Rust.

## Ready?

<!-- If you feel like you can help to check off the following tasks, that'd be great. If not, don't worry - we will take care of it. -->

- [x] Added unit tests for fixed bug/feature
- [x] Passing all unit tests
- [x] Complete build `./mvnw clean verify` passes (checked automatically by github actions)
- [x] Added (in-code) documentation (if needed)

adangel commented on pmd-rust/src/main/java/net/sourceforge/pmd/lang/rust/RustLanguageModule.java:
```diff
@@ -0,0 +1,29 @@
+    private static final String ID = "rust";
+
+    public RustLanguageModule() {
+        super(LanguageMetadata.withId(ID).name("Rust").extensions("rust"));
```
```suggestion
        super(LanguageMetadata.withId(ID).name("Rust").extensions("rs"));
```

juliapaluch replied:
thank you! addressed in da524115c0b943256a1466c4d0145b0fc49b1a8f

juliapaluch commented:
good point! I added a README to the directory where the grammar lives. I also deleted `RustParser.g4` because I'm assuming that's unnecessary if we're just tokenizing. Please correct me if I'm wrong and I'll add it back!

jsotuyod commented on pmd-rust/src/main/java/net/sourceforge/pmd/lang/rust/ast/RustLexer.java:
```diff
@@ -0,0 +1,1233 @@
+import net.sourceforge.pmd.lang.rust.ast.RustLexerBase;
+
+@SuppressWarnings({"all", "warnings", "unchecked", "unused", "cast", "CheckReturnValue", "this-escape"})
+public class RustLexer extends RustLexerBase {
```
why is this file here? if it's generated from the grammar during build (under `build/generated-sources/antlr4`) there is no need to have it under `src/main/java`…

juliapaluch replied:
oops! that's my bad. I inlined the custom functions from the ANTLR grammar in `RustLexer.g4` instead and added a note about the modification to the README in 0f58dae1f25ae3c6b72f6dc1e070d86599c3e716. thank you! :)

jsotuyod commented on pmd-rust/src/main/java/net/sourceforge/pmd/lang/rust/ast/RustLexerBase.java:
```diff
@@ -0,0 +1,84 @@
+
+import org.antlr.v4.runtime.*;
+
+public abstract class RustLexerBase extends Lexer{
```
ditto

pmd-test commented:
<!--
  0 Errors
  0 Warnings
  1 Message: No regression tested rules hav...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">No regression tested rules have been changed.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

adangel reviewed (approved):
Thanks!
